vector-im/element-android
# Implement logic for sign in with QR

This PR is dependent on https://github.com/vector-im/element-android/pull/7338. [Compare](https://github.com/vector-im/element-android/compare/feature/ons/qr_code_login_ui...feature/hughns/qr_code_login)

## Type of change


## Content



This PR has the logic to implement the UI from https://github.com/vector-im/element-android/pull/7338

## Motivation and context

The motivation is as per https://github.com/vector-im/element-android/pull/7338



## Screenshots / GIFs

See https://github.com/vector-im/element-android/pull/7338





## Tests



Testing requires generating a QR code in Element Web and scanning it in Element Android.

To generate a code in Element Web:

- Go to preview build of Element Web: https://pr9403--matrix-react-sdk.netlify.app/
- Choose test homeserver: synapse-msc3882.rendezvous.lab.element.dev
- Register a user
- Go to Menu -> All settings -> Labs and enable the QR feature

<img width="952" alt="image" src="https://user-images.githubusercontent.com/6955675/196170034-d2aa3563-cabe-4dff-9a03-e9cda4ea337f.png">

- Then go to Security & Privacy and see the new Sign in with QR section:
<img width="1003" alt="image" src="https://user-images.githubusercontent.com/6955675/196170177-668312ab-d54a-4293-97a9-c24b7e32aeca.png">

- Press Show QR to generate a QR:
<img width="650" alt="image" src="https://user-images.githubusercontent.com/6955675/196170219-65a9353c-222e-4fda-be6b-d07e3e838485.png">


To scan in Android:

- Launch app when signed out
- Enable the QR Code Login feature in Features:
<img width="436" alt="image" src="https://user-images.githubusercontent.com/6955675/196170493-17101b06-14d9-4966-b580-d54300eddecb.png">
- Go back to main screen
- Select "I already have an account"
- Press EDIT to enter a different homeserver
- Enter https://synapse-msc3882.rendezvous.lab.element.dev
- Press next
- You should then see a new button:
<img width="423" alt="image" src="https://user-images.githubusercontent.com/6955675/196170931-da8b5d9d-0198-45ca-b157-d82c407fd8fa.png">
- Press Scan QR code and follow instructions

## Tested devices

- OS version(s):

## Checklist




Signed-off-by: Hugh Nimmo-Smith <hughns@element.io>
